plugins/civo/api_key.go

```diff
@@ -0,0 +1,138 @@
+				MarkdownDescription: "API Key used to authenticate to Civo.",
+				Secret:              true,
+				Composition: &schema.ValueComposition{
+					Length: 50,
+					Charset: schema.Charset{
+						Uppercase: true,
+						Lowercase: true,
+						Digits:    true,
+					},
+				},
+			},
+			{
+				Name:                fieldname.APIKeyID,
+				MarkdownDescription: "The Name of apikey used to authenticate to civo",
+				Optional:            true,
+			},
+		},
+		DefaultProvisioner: civoProvisioner{},
+		Importer: importer.TryAll(
+			importer.TryEnvVarPair(defaultEnvVarMapping),
+			importer.TryEnvVarPair(secondEnvVarMapping),
+			TryCivoConfigFile("~/.civo.json"),
+		)}
+}
+
+var defaultEnvVarMapping = map[string]sdk.FieldName{
+	"CIVO_TOKEN": fieldname.APIKey,
+}
+
+var secondEnvVarMapping = map[string]sdk.FieldName{
+	"CIVO_API_KEY":      fieldname.APIKey,
+	"CIVO_API_KEY_NAME": fieldname.APIKeyID,
+}
+
+type civoProvisioner struct {
+	sdk.Provisioner
+	Schema map[string]sdk.FieldName
+}
+
+func (p civoProvisioner) Provision(ctx context.Context, in sdk.ProvisionInput, out *sdk.ProvisionOutput) {
+	if apiKey, ok := in.ItemFields[fieldname.APIKey]; ok {
+		out.AddEnvVar("CIVO_TOKEN", apiKey)
+	}
+
+	if apiKey, ok := in.ItemFields[fieldname.APIKey]; ok {
+		out.AddEnvVar("CIVO_API_KEY", apiKey)
+		//out.AddEnvVar("CIVO_API_KEY_NAME", apiKeyID)
+	}
+
+	if apiKeyID, ok := in.ItemFields[fieldname.APIKeyID]; ok {
+		out.AddEnvVar("CIVO_API_KEY_NAME", apiKeyID)
+	}
+}
+
+func (p civoProvisioner) Deprovision(ctx context.Context, in sdk.DeprovisionInput, out *sdk.DeprovisionOutput) {
+	// Nothing to do here: environment variables get wiped automatically when the process exits.
+}
+
+func (p civoProvisioner) Description() string {
+	var envVarNames []string
+	for envVarName := range p.Schema {
+		envVarNames = append(envVarNames, envVarName)
+	}
+	return strings.Join(envVarNames, ", ")
+}
+
+func TryCivoConfigFile(path string) sdk.Importer {
+
+	return importer.TryFile(path, func(ctx context.Context, contents importer.FileContents, in sdk.ImportInput, out *sdk.ImportAttempt) {
+		var config Config
+		if err := contents.ToJSON(&config); err != nil {
+			out.AddError(err)
+			return
+
+		}
+
+		if len(config.APIKeys) == 0 {
+			return
+		}
+
+		for key := range config.APIKeys {
+			apiKey := os.Getenv("CIVO_TOKEN")
+			if apiKey == "" {
+				apiKey = os.Getenv("CIVO_API_KEY")
+				if apiKey == "" {
+					continue
+				}
+
+				apiKeyName := os.Getenv("CIVO_API_KEY_NAME")
+				if apiKeyName == "" {
+					continue
+				}
+
+				out.AddCandidate(sdk.ImportCandidate{
+					Fields: map[sdk.FieldName]string{
+						fieldname.APIKey:   apiKey,
+						fieldname.APIKeyID: apiKeyName,
+					},
+					NameHint: importer.SanitizeNameHint(key),
+				})
+			} else {
+				out.AddCandidate(sdk.ImportCandidate{
+					Fields: map[sdk.FieldName]string{
+						fieldname.APIKey: apiKey,
+					},
+					NameHint: importer.SanitizeNameHint(key),
+				})
+			}
+		}
```

Comment:
Could we stick to only importing only from the configuration file in the `TryFile` block, and not from the environment variables? Importing from environment variables is already covered in the `TryEnvVarPair` further above.

Reply:
Okay